File: kafka-reassign-tool.rb
#!/usr/bin/ruby

require 'yaml'
require 'json'
require 'open3'
require 'tempfile'
require 'optparse'

DEFAULT_KAFKA_ROOT='/usr/share/varadhi-kafka'
DEFAULT_KAFKA_CONFIG='/config/server.properties'

$kafka_root = nil
$zookeeper_url = nil
$input_assignment = nil
$throttle = nil
$retry_after = nil
$debug_log = false

def run(command, args = [], opts = {})

  opts = {
      :raise_on_exitstatus => true,
      :raise_on_err => true,
      :input => ''
  }.merge(opts)

  out = nil
  err = nil
  exit_status = nil
  Open3::popen3(command, *args) do |stdin, stdout, stderr, wait_thr|
    stdin << opts[:input]
    stdin.close
    out = stdout.read
    err = stderr.read
    exit_status = wait_thr.value
  end

  if (opts[:raise_on_exitstatus] && exit_status.exitstatus != 0) || (opts[:raise_on_err] && !err.empty?) then
    puts "#{command} failed. exitstatus=#{exit_status.exitstatus}, stderr:"
    puts err
    raise "#{command} failed"
  end

  {:out => out, :err => err, :exit_status => exit_status}
end

def get_kafka_root()
  return $kafka_root if $kafka_root
  DEFAULT_KAFKA_ROOT
end

def get_zk_url()
  return $zookeeper_url if $zookeeper_url

  config_file = get_kafka_root + DEFAULT_KAFKA_CONFIG

  if File.file? config_file then
    puts "Reading #{config_file}"
    File.open(config_file, 'r') do |file|
      file.each_line do |line|
        if line =~ /^zookeeper.connect=(.*)/ then
          $zookeeper_url = $1.strip
        end
      end
    end
  else
    puts "Config file #{config_file} does not exist"
  end

  raise "No zookeeper URL given" unless $zookeeper_url

  $zookeeper_url
end

def create_assignment_json(topic, partition, brokers)
  data = {
      :partitions => [{
                          :topic => topic,
                          :partition => partition,
                          :replicas => brokers
                      }],
      :version => 1
  }

  assignment_file = Tempfile.new("#{topic}-#{partition}")
  assignment_file << data.to_json()
  assignment_file.close

  puts "temp file: #{assignment_file.path}. content: #{data.to_json}" if $debug_log

  assignment_file.path
end

def verify_assignment(assignment_file)
  result = run(get_kafka_root + '/bin/kafka-reassign-partitions.sh', [
      '--zookeeper', get_zk_url,
      '--reassignment-json-file', assignment_file,
      '--verify'
  ])

  reassignment_lines = result[:out].lines.collect(&:strip).select { |l| l.start_with? "Reassignment" }
  in_progress = reassignment_lines.select { |l| l.end_with? "is still in progress" }.length
  completed = reassignment_lines.select { |l| l.end_with? "completed successfully" }.length

  puts "verify output: #{result[:out]}in_progress: #{in_progress}. completed: #{completed}" if $debug_log

  {
      :in_progress => in_progress,
      :completed => completed,
      :out => result[:out],
      :err => result[:err]
  }
end

def begin_reassignment(assignment_file)
  result = run(get_kafka_root + '/bin/kafka-reassign-partitions.sh', [
      '--zookeeper', get_zk_url,
      '--reassignment-json-file', assignment_file,
      '--execute',
      '--throttle', $throttle
  ])

  msg = result[:out].lines.collect(&:strip).select { |l| l.to_s.include? "Successfully started reassignment of partitions" }
  started = msg.length == 1

  puts "begin output: #{result[:out]}started: #{started}" if $debug_log

  if !started then
    puts "couldn't find the started message when starting the reassignment."
    puts "tool output:\n#{result[:out]}"
    puts "\njson_file:\n#{File.read(assignment_file)}"
    raise 'couldnt start the reassignment'
  end
end

def reassign(topic, partition, brokers)
  # first verify that it has already been reassigned.
  file_path = create_assignment_json(topic, partition, brokers)

  progress = verify_assignment file_path

  if (progress[:completed] == 1) then
    puts "#{topic}-#{partition} completed successfully"
    return
  end

  if (progress[:in_progress] == 0) then
    puts "#{topic}-#{partition} starting reassignment of partitions"

    begin_reassignment file_path

    puts "#{topic}-#{partition} started reassignment of partitions"
    sleep 5
  end

  loop do
    progress = verify_assignment file_path
    if (progress[:completed] == 1) then
      puts "#{topic}-#{partition} completed successfully"
      return
    elsif (progress[:in_progress] == 1) then
      puts "#{topic}-#{partition} is still in progress. Retrying after #{$retry_after} seconds"
    else
      puts "#{topic}-#{partition} failed.\nout:\n#{progress[:out]}\nerr:\n#{progress[:err]}"
      return
    end

    sleep $retry_after
  end
end

optparse = OptionParser.new do |opts|
  opts.on("--kafka-home <dir>",
          "Root directory of the Kafka installation",
          "  (standard Kafka scripts must be under bin/ directory there)",
          "  Default: #{DEFAULT_KAFKA_ROOT}") { |v| $kafka_root = v }
  opts.on("--zookeeper <url>",
          "The connection string for the zookeeper connection",
          "  If not specified, and attempt to read it from Kafka config file is made") { |v| $zookeeper_url = v }
  opts.on("--input <file>",
          "File containing partition assignment") { |v| $input_assignment = JSON.parse(File.read(v)) }
  opts.on("--throttle <throttle>",
          "replication throttle in B/s") { |v| $throttle = v }
  opts.on("--retry-after <retry>",
          "retry duration in sec after which the tool should look for completion status again") { |v| $retry_after = v.to_i }
  opts.on("--debug",
          "for debug logs") { |_| $debug_log = true }
end

begin
  optparse.parse!
  raise OptionParser::MissingArgument.new('input') if $input_assignment.nil?
  raise OptionParser::MissingArgument.new('throttle') if $throttle.nil?
  raise OptionParser::MissingArgument.new('retry-after') if $retry_after.nil?
rescue OptionParser::InvalidOption, OptionParser::MissingArgument
  puts $!.to_s
  puts optparse
  exit -1
end


zk_url = get_zk_url()
puts "Using zookeeper URL: #{zk_url}"

$input_assignment.each do |assignment|
  topic = assignment["topic"]
  partition = assignment["partition"]
  brokers = assignment["to"]

  reassign(topic, partition, brokers)
end

File: kafka-reassign-tool.py
import os
import os.path
import json
import subprocess
import tempfile
import argparse
import logging
import re
import time

DEFAULT_KAFKA_ROOT = '/usr/share/varadhi-kafka'
DEFAULT_KAFKA_CONFIG = '/config/server.properties'
DEFAULT_RETRY_AFTER = 60

input_file = None
kafka_root = None
zookeeper_url = None
input_assignment = None
throttle = None
throttle_getter = None
retry_after = None


def set_logger(debug):
    level = logging.DEBUG if debug else logging.INFO
    logging.basicConfig(format='%(asctime)s %(message)s', level=level)


def output_to_lines(output):
    if output == None or output.strip() == "":
        return None
    return [x.strip() for x in output.splitlines()]


def run(command, args=[], check=False):
    cp = subprocess.run([command] + args, universal_newlines=True,
                        stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=check)
    return (output_to_lines(cp.stdout), output_to_lines(cp.stderr), cp.returncode)


def get_kafka_root():
    return kafka_root


def get_zk_url():
    global zookeeper_url

    if zookeeper_url is not None:
        return zookeeper_url

    config_file = get_kafka_root() + DEFAULT_KAFKA_CONFIG

    if os.path.isfile(config_file):
        logging.info("Reading %s", config_file)
        with open(config_file, 'r') as f:
            for l in f.read().splitlines():
                m = re.search("^zookeeper.connect=(.*)", l)
                if m is not None:
                    zookeeper_url = m.groups()[0]
                    break
    else:
        logging.info("Config file %s does not exist", config_file)

    if zookeeper_url is None:
        raise Exception("No zookeeper URL given")

    return zookeeper_url


def get_temp_file_name(id):
    with tempfile.NamedTemporaryFile(prefix=input_file + "-" + str(id) + "-", delete=False) as tmp:
        return tmp.name


def create_assignment_json(assignments, id):
    data = {
        "partitions": [
            {"topic": x["topic"], "partition": x["partition"], "replicas": x["to"]} for x in assignments
        ],
        "version": 1
    }
    file_path = get_temp_file_name(id)
    with open(file_path, 'w') as f:
        json.dump(data, f, indent=2)
    logging.info("temp file: %s", file_path)
    logging.info("content: %s", json.dumps(data))
    return file_path


def verify_assignment(assignment_file):
    (stdout, stderr, code) = run(get_kafka_root() + '/bin/kafka-reassign-partitions.sh',
                                 ['--zookeeper', get_zk_url(), '--reassignment-json-file', assignment_file, '--verify'])
    if code != 0:
        raise Exception("error while verifying\ncode: {}\nstderr: {}".format(code, stderr))
    reassignment_lines = [x for x in stdout if x.startswith("Reassignment")]
    in_progress = len([x for x in reassignment_lines if x.endswith("is still in progress")])
    completed = len([x for x in reassignment_lines if x.endswith("completed successfully")])
    logging.debug("verify output: %s", "\n".join(stdout))
    logging.debug("in_progress: %s", in_progress)
    logging.debug("completed: %s", completed)
    return (in_progress, completed, stdout, stderr)


def begin_reassignment(assignment_file, new_throttle):
    (stdout, stderr, code) = run(get_kafka_root() + '/bin/kafka-reassign-partitions.sh',
                                ['--zookeeper', get_zk_url(), '--reassignment-json-file', assignment_file, '--execute', '--throttle', str(new_throttle)])
    if code != 0:
        raise Exception("error while begin assign\ncode: {}\nstderr: {}".format(code, stderr))
    msg = [x for x in stdout if "Successfully started reassignment of partitions" in x]
    started = True if len(msg) == 1 else False
    logging.debug("begin output: \n%s\nstarted: %s", "\n".join(stdout), started)
    if not started:
        logging.info("couldn't find the started message when starting the reassignment.")
        logging.info("tool output:\n%s", stdout)
        raise Exception("couldnt start the reassignment")


def change_throttle(assignment_file, new_throttle):
    (stdout, stderr, code) = run(get_kafka_root() + '/bin/kafka-reassign-partitions.sh',
                                ['--zookeeper', get_zk_url(), '--reassignment-json-file', assignment_file, '--execute', '--throttle', str(new_throttle)])
    if code != 0:
        raise Exception("error while begin assign\ncode: {}\nstderr: {}".format(code, stderr))
    msg = [x for x in stdout if "There is an existing assignment running" in x]
    changed = True if len(msg) == 1 else False
    logging.debug("change throttle output: \n%s\changed: %s", "\n".join(stdout), changed)
    if not changed:
        logging.info("couldn't find the existing assignment running msg while changing throttle.")
        logging.info("tool output:\n%s", stdout)
        raise Exception("couldnt change the throttle")


def next_throttle(i):
    return (throttle[i:-1] + throttle[-1:])[0]

def reassign(assignments, id):
    # first verify that it has already been reassigned.
    file_path = create_assignment_json(assignments, id)
    (in_progress, completed, stdout, stderr) = verify_assignment(file_path)
    partitions = len(assignments)
    if completed == partitions:
        logging.info("reassignment completed successfully")
        return
    if in_progress == 0:
        logging.info("starting reassignment of partitions")
        begin_reassignment(file_path, next_throttle(0))
        logging.info("started reassignment of partitions")

    retry_count = 0
    while True:
        time.sleep(retry_after)
        (in_progress, completed, stdout, stderr) = verify_assignment(file_path)
        if completed == partitions:
            logging.info("reassignment completed successfully")
            return
        elif in_progress > 0:
            logging.info("reassignemnt is still in progress. Retrying after %s seconds", retry_after)
        else:
            raise Exception("reassignment failed.\nstdout:\n%s\nstderr:\n%s", "\n".join(stdout), "\n".join("stderr"))
        retry_count = retry_count + 1
        if retry_count % 2 == 0 and retry_count < (2 * len(throttle)):
            iteration = retry_count // 2
            new_throttle = next_throttle(iteration)
            logging.info("changing throttle to: %s", new_throttle)
            change_throttle(file_path, new_throttle)
            logging.info("changed throttle successfully")


def load_throttle_from_file():
    with open('throttle.json', 'r') as f:
        return json.load(f)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--kafka-home", help="Root directory of the Kafka installation. Default: {}".format(
        DEFAULT_KAFKA_ROOT), default=DEFAULT_KAFKA_ROOT)
    parser.add_argument(
        "--zookeeper", help="The connection string for the zookeeper connection. If not specified, an attempt to read it from Kafka config file is made")
    parser.add_argument("input", help="File containing partition assignment")
    parser.add_argument(
        "--throttle", help="Replication throttle in B/s. If not given, throttle.json file will be loaded", type=str, default=None)
    parser.add_argument("--retry-after", help="Retry duration in sec after which the tool should look for completion status again",
                        type=int, default=DEFAULT_RETRY_AFTER)
    parser.add_argument("--debug", help="For debug logs", action="store_true")

    args = parser.parse_args()

    kafka_root = args.kafka_home
    zookeeper_url = args.zookeeper
    input_file = args.input
    with open(args.input, 'r') as f:
        input_assignment = json.load(f)
    retry_after = args.retry_after
    set_logger(args.debug)

    if(args.throttle is None):
        throttle_getter = load_throttle_from_file
    else:
        t = [int(x) for x in args.throttle.split(",")]
        throttle_getter = lambda : t

    throttle = throttle_getter()

    logging.info("Using:")
    logging.info("kafka root: %s", kafka_root)
    logging.info("zk url: %s", get_zk_url())
    logging.info("throttle: %s", throttle)
    logging.info("first throttle: %s", next_throttle(0))
    logging.info("retry after: %s", retry_after)



    for i in range(0, len(input_assignment)):
        # refetch throttle values
        throttle = throttle_getter()
        reassign(input_assignment[i], i)
